File: apps/backend-python/src/housing_platform/payments/repository.py
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

from sqlalchemy import select
from sqlalchemy.orm import Session

from housing_platform.auth.errors import (
    BadRequestError,
    BookingExpiredError,
    ForbiddenError,
)
from housing_platform.db.models import (
    BookingPriceSnapshots,
    Bookings,
    Payments,
    PlatformSettings,
    Properties,
)


@dataclass(frozen=True)
class PaymentOrderRow:
    payment_id: uuid.UUID
    order_id: uuid.UUID
    booking_id: uuid.UUID
    amount_krw: int
    order_name: str


@dataclass(frozen=True)
class PaymentLookupRow:
    id: uuid.UUID
    order_id: uuid.UUID
    booking_id: uuid.UUID
    customer_id: uuid.UUID
    amount_krw: int
    status: str


class PaymentRepository:
    def __init__(self, db: Session) -> None:
        self._db = db

    def get_payment_by_order_id(self, order_id: uuid.UUID) -> PaymentLookupRow | None:
        payment = self._db.scalar(select(Payments).where(Payments.order_id == order_id))
        if payment is None:
            return None
        return PaymentLookupRow(
            id=payment.id,
            order_id=payment.order_id,
            booking_id=payment.booking_id,
            customer_id=payment.customer_id,
            amount_krw=payment.amount_krw,
            status=payment.status,
        )

    def create_payment_order(
        self,
        booking_id: uuid.UUID,
        customer_id: uuid.UUID,
    ) -> PaymentOrderRow:
        booking = self._db.scalar(
            select(Bookings).where(Bookings.id == booking_id).with_for_update()
        )
        if booking is None or booking.customer_id != customer_id:
            raise ForbiddenError("Booking not found")

        if booking.status == "payment_failed":
            if booking.payment_retry_count >= 1:
                raise BadRequestError("Payment retry limit reached")

            settings = self._get_platform_settings()
            booking.status = "pending_payment"
            booking.payment_retry_count = booking.payment_retry_count + 1
            booking.hold_expires_at = datetime.now(UTC) + timedelta(
                minutes=settings.hold_ttl_minutes
            )
            self._db.flush()
        elif booking.status != "pending_payment":
            raise ForbiddenError("Booking is not awaiting payment")

        if booking.hold_expires_at is not None and booking.hold_expires_at <= datetime.now(UTC):
            booking.status = "expired"
            self._db.flush()
            raise BookingExpiredError("Booking hold has expired")

        snapshot = self._db.scalar(
            select(BookingPriceSnapshots).where(BookingPriceSnapshots.booking_id == booking_id)
        )
        if snapshot is None:
            raise BadRequestError("Booking price snapshot missing")

        property_title = self._db.scalar(
            select(Properties.title).where(Properties.id == booking.property_id)
        )

        order_id = uuid.uuid4()
        payment = Payments(
            order_id=order_id,
            booking_id=booking_id,
            customer_id=customer_id,
            amount_krw=snapshot.total_krw,
            status="pending",
        )
        self._db.add(payment)
        self._db.flush()

        return PaymentOrderRow(
            payment_id=payment.id,
            order_id=payment.order_id,
            booking_id=payment.booking_id,
            amount_krw=payment.amount_krw,
            order_name=property_title or "Housing Platform stay",
        )

    def _get_platform_settings(self) -> PlatformSettings:
        settings = self._db.scalar(select(PlatformSettings).where(PlatformSettings.id == 1))
        if settings is None:
            raise BadRequestError("Platform settings are not configured")
        return settings

File: apps/backend-go/internal/payments/repository.go
package payments

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/jackc/pgx/v5/pgxpool"

	apierrors "github.com/housing-platform/backend-go/internal/api/errors"
	"github.com/housing-platform/backend-go/internal/db"
	"github.com/housing-platform/backend-go/internal/shared"
)

type paymentLookupRow struct {
	ID         pgtype.UUID
	OrderID    pgtype.UUID
	BookingID  pgtype.UUID
	CustomerID pgtype.UUID
	AmountKrw  int32
	Status     string
}

type paymentOrderRow struct {
	PaymentID pgtype.UUID
	OrderID   pgtype.UUID
	BookingID pgtype.UUID
	AmountKrw int32
	OrderName string
}

type Repository struct {
	pool *pgxpool.Pool
}

func NewRepository(pool *pgxpool.Pool) *Repository {
	return &Repository{pool: pool}
}

func (r *Repository) GetPaymentByOrderID(ctx context.Context, q pgx.Tx, orderID pgtype.UUID) (*paymentLookupRow, error) {
	var row paymentLookupRow
	err := q.QueryRow(ctx, `
		SELECT id, order_id, booking_id, customer_id, amount_krw, status::text
		FROM public.payments WHERE order_id = $1
	`, orderID).Scan(&row.ID, &row.OrderID, &row.BookingID, &row.CustomerID, &row.AmountKrw, &row.Status)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	return &row, nil
}

func (r *Repository) CreatePaymentOrder(ctx context.Context, q pgx.Tx, bookingID, customerID pgtype.UUID) (paymentOrderRow, error) {
	var booking db.Booking
	err := q.QueryRow(ctx, `
		SELECT id, customer_id, status, hold_expires_at, payment_retry_count, property_id
		FROM public.bookings WHERE id = $1 FOR UPDATE
	`, bookingID).Scan(&booking.ID, &booking.CustomerID, &booking.Status, &booking.HoldExpiresAt, &booking.PaymentRetryCount, &booking.PropertyID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			panic(apierrors.Forbidden("Booking not found"))
		}
		return paymentOrderRow{}, err
	}
	if shared.UUIDToString(booking.CustomerID) != shared.UUIDToString(customerID) {
		panic(apierrors.Forbidden("Booking not found"))
	}

	now := time.Now().UTC()

	if booking.Status == db.BookingStatusPaymentFailed {
		if booking.PaymentRetryCount >= 1 {
			panic(apierrors.BadRequest("Payment retry limit reached"))
		}
		holdTTL, err := shared.NewPricingService(r.pool).GetHoldTTLMinutes(ctx, q)
		if err != nil {
			return paymentOrderRow{}, err
		}
		holdExpires := pgtype.Timestamptz{Time: now.Add(time.Duration(holdTTL) * time.Minute), Valid: true}
		_, err = q.Exec(ctx, `
			UPDATE public.bookings
			SET status = 'pending_payment', payment_retry_count = payment_retry_count + 1, hold_expires_at = $2
			WHERE id = $1
		`, bookingID, holdExpires)
		if err != nil {
			return paymentOrderRow{}, err
		}
	} else if booking.Status != db.BookingStatusPendingPayment {
		panic(apierrors.Forbidden("Booking is not awaiting payment"))
	}

	if booking.HoldExpiresAt.Valid && !booking.HoldExpiresAt.Time.After(now) {
		_, _ = q.Exec(ctx, `UPDATE public.bookings SET status = 'expired' WHERE id = $1`, bookingID)
		panic(apierrors.BookingExpired("Booking hold has expired"))
	}

	var totalKrw int32
	err = q.QueryRow(ctx, `
		SELECT total_krw FROM public.booking_price_snapshots WHERE booking_id = $1
	`, bookingID).Scan(&totalKrw)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			panic(apierrors.BadRequest("Booking price snapshot missing"))
		}
		return paymentOrderRow{}, err
	}

	var propertyTitle *string
	_ = q.QueryRow(ctx, `SELECT title FROM public.properties WHERE id = $1`, booking.PropertyID).Scan(&propertyTitle)

	orderName := "Housing Platform stay"
	if propertyTitle != nil && *propertyTitle != "" {
		orderName = *propertyTitle
	}

	orderUUID := uuid.New()
	var paymentID pgtype.UUID
	err = q.QueryRow(ctx, `
		INSERT INTO public.payments (order_id, booking_id, customer_id, amount_krw, status)
		VALUES ($1, $2, $3, $4, 'pending')
		RETURNING id
	`, orderUUID, bookingID, customerID, totalKrw).Scan(&paymentID)
	if err != nil {
		return paymentOrderRow{}, err
	}

	var orderID pgtype.UUID
	_ = orderID.Scan(orderUUID.String())

	return paymentOrderRow{
		PaymentID: paymentID,
		OrderID:   orderID,
		BookingID: bookingID,
		AmountKrw: totalKrw,
		OrderName: orderName,
	}, nil
}
